agents/utils/helpers.py

```diff
@@ -0,0 +1,87 @@
+import logging
+import re
+import subprocess
+import time
+
+from .logtools import maybe_log_message
+
+
+def restart_service(logger, fallback_logger, service, attempts=3):
+    """
+    Attempts to restart a system service with exponential backoff
+    if it is found to be inactive. Logs each attempt and result.
+
+    Parameters:
+        agent (ServerAgent): Agent instance used for logging. Must have
+        `logger` and `fallback_logger` attributes.
+        service (str): Name of the system service to restart (e.g., 'ssh').
+
+    Returns:
+        bool: True if the service was restarted, False otherwise.
+
+    Notes:
+        This function uses 'sudo systemctl restart <service>' and expects
+        that the agent has sufficient privileges to perform the operation.
+        Backoff strategy uses 2s, 4s, and 8s delays between attempts.
+    """
+
+    maybe_log_message(
+        '%s not active. Attempting restart...' % service,
+        logger,
+        fallback_logger=fallback_logger
+        )
+
+    for i in range(1, attempts+1):
+        try:
+            delay = 2**i
+
+            maybe_log_message(
+                'Restarting %s (delay before restart: %s).' % (service, delay),
+                logger,
+                fallback_logger=fallback_logger
+                )
```

Comment:
It could be helpful for debugging to log which attempt is being made:
```
maybe_log_message(
    'Attempt #%d: Restarting %s (delay: %ss)' % (i, service, delay),
    ...
)
```